### PR TITLE
FEATURE: new strategy dca2 perparation

### PR DESCRIPTION
**Pull request by kbearXD**

### CODE REVIEW

File: pkg/strategy/dca2/strategy.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+func (s *Strategy) ID() string {
+	return ID
+}
+
+func (s *Strategy) Validate() error {
+	if s.OrderNum < 1 {
+		return fmt.Errorf("maxOrderNum can not be < 1")
+	}
+
+	if s.TakeProfitSpread.Compare(fixedpoint.Zero) <= 0 {
```
Comment (narumiruna) [inline pkg/strategy/dca2/strategy.go:62]: ```s.TakeProfitSpread.Float64() <= 0``` or ```s.TakeProfitSpread.Sign() <= 0```

File: pkg/strategy/dca2/strategy.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+	instanceID := s.InstanceID()
+	s.session = session
+	s.orderExecutor = bbgo.NewGeneralOrderExecutor(session, s.Symbol, ID, instanceID, s.Position)
+	s.orderExecutor.BindEnvironment(s.Environment)
+	s.orderExecutor.BindProfitStats(s.ProfitStats)
+	s.orderExecutor.TradeCollector().OnPositionUpdate(func(position *types.Position) {
+		bbgo.Sync(ctx, s)
+	})
+	s.orderExecutor.Bind()
```
Comment (narumiruna) [inline pkg/strategy/dca2/strategy.go:114]: https://github.com/c9s/bbgo/blob/580c6d2030c2badb8c16040f000963fe9281879a/pkg/strategy/rsicross/strategy.go#L52-L53
You can refactor the code using `common.Strategy`

File: pkg/strategy/dca2/strategy.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+type Strategy struct {
+	Environment *bbgo.Environment
+	Market      types.Market
+
+	Symbol string `json:"symbol"`
+
+	// setting
+	Budget           fixedpoint.Value `json:"budget"`
+	OrderNum         int64            `json:"orderNum"`
+	Margin           fixedpoint.Value `json:"margin"`
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:32]: what's the Margin?

File: pkg/strategy/dca2/strategy.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	TakeProfitSpread fixedpoint.Value `json:"takeProfitSpread"`
+	RoundInterval    types.Duration   `json:"roundInterval"`
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:34]: should be CoolDownInterval

File: pkg/strategy/dca2/strategy.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	Environment *bbgo.Environment
+	Market      types.Market
+
+	Symbol string `json:"symbol"`
+
+	// setting
+	Budget           fixedpoint.Value `json:"budget"`
+	OrderNum         int64            `json:"orderNum"`
+	Margin           fixedpoint.Value `json:"margin"`
+	TakeProfitSpread fixedpoint.Value `json:"takeProfitSpread"`
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:33]: TakeProfitRatio

File: pkg/strategy/dca2/strategy.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	s.orderExecutor.BindEnvironment(s.Environment)
+	s.orderExecutor.BindProfitStats(s.ProfitStats)
+	s.orderExecutor.TradeCollector().OnPositionUpdate(func(position *types.Position) {
+		bbgo.Sync(ctx, s)
+	})
+	s.orderExecutor.Bind()
+	s.book = types.NewStreamBook(s.Symbol)
+	s.book.BindStream(s.session.MarketDataStream)
+
+	balances := session.GetAccount().Balances()
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:118]: better to use "QueryBalances" here

File: pkg/strategy/dca2/strategy.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	balance := balances[s.Market.QuoteCurrency]
+	if balance.Available.Compare(s.Budget) < 0 {
+		return fmt.Errorf("the available balance of %s is %s which is less than budget setting %s, please check it", s.Market.QuoteCurrency, balance.Available, s.Budget)
+	}
+
+	session.MarketDataStream.OnBookUpdate(func(book types.SliceOrderBook) {
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:124]: Avoid subscribing order book, which consumes too much traffic

File: pkg/strategy/dca2/strategy_test.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+import (
+	"testing"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+func number(a interface{}) fixedpoint.Value {
```
Comment (c9s) [inline pkg/strategy/dca2/strategy_test.go:12]: you can dot import the testhelper

**Review by gx578007 — APPROVED** [review]

File: pkg/strategy/dca2/strategy.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+	balances, err := session.Exchange.QueryAccountBalances(ctx)
+	if err != nil {
+		return err
+	}
+
+	balance := balances[s.Market.QuoteCurrency]
+	if balance.Available.Compare(s.Budget) < 0 {
+		return fmt.Errorf("the available balance of %s is %s which is less than budget setting %s, please check it", s.Market.QuoteCurrency, balance.Available, s.Budget)
+	}
```
Comment (c9s) [inline pkg/strategy/dca2/strategy.go:145]: you need to do the checks before you add the callbacks

**Pull request merged**